crypto/hash/sha3.go

```diff
@@ -1,250 +1,51 @@
-	s.Reset()
-	s.write(data)
-	return s.sum()
-}
-
-// SumHash returns the SHA3-256 digest of the data written to the state.
-// It updates the state and doesn't allow further writing without
-// calling Reset().
-func (s *sha3State) SumHash() Hash {
-	return s.sum()
-}
-
-// Write absorbs more data into the hash's state.
-// It returns the number of bytes written and never errors.
-func (d *sha3State) Write(p []byte) (int, error) {
-	d.write(p)
-	return len(p), nil
-}
-
 // ComputeSHA3_256 computes the SHA3-256 digest of data
 // and copies the result to the result buffer.
 //
 // The function is not part of the Hasher API. It is a light API
 // that allows a simple computation of a hash and minimizes
 // heap allocations.
-func ComputeSHA3_256(result *[HashLenSha3_256]byte, data []byte) {
-	state := &sha3State{
-		rate:      rateSha3_256,
-		outputLen: HashLenSha3_256,
+func ComputeSHA3_256(result *[HashLenSHA3_256]byte, data []byte) {
+	state := &spongeState{
+		rate:      rateSHA3_256,
+		dsByte:    dsByteSHA3,
+		outputLen: HashLenSHA3_256,
 		bufIndex:  bufNilValue,
 		bufSize:   bufNilValue,
 	}
 	state.write(data)
 	state.padAndPermute()
 	copyOut(result[:], state)
 }
-
-// The functions below were copied and modified from golang.org/x/crypto/sha3.
-//
-// Copyright (c) 2009 The Go Authors. All rights reserved.
-
-// Redistribution and use in source and binary forms, with or without
-// modification, are permitted provided that the following conditions are
-// met:
-
-//    * Redistributions of source code must retain the above copyright
-// notice, this list of conditions and the following disclaimer.
-//    * Redistributions in binary form must reproduce the above
-// copyright notice, this list of conditions and the following disclaimer
-// in the documentation and/or other materials provided with the
-// distribution.
-//    * Neither the name of Google Inc. nor the names of its
-// contributors may be used to endorse or promote products derived from
-// this software without specific prior written permission.
-
-// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
-// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
-// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
-// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
-// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
-// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
-// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
-// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
-// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
-// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
-// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-
-const (
-	// maxRate is the maximum size of the internal buffer. SHA3-256
-	// currently needs the largest buffer.
-	maxRate = 1088 / 8
```

Comment:
I guess we don't need this one anymore? do we have similar logic somewhere else ?
??

Reply:
The same constant just moved here : https://github.com/onflow/flow-go/blob/tarak/6106-keccak-256/crypto/hash/keccak.go#L70